### PR TITLE
Reader: preserve linebreaks for excerpts on summary-only feeds

### LINKED ISSUE #696: Reader: excerpts in summary feeds don't display line breaks
Steps to reproduce:

1. Go to WP Admin > Settings > Reading, select "Summary" for feeds, and Save Changes
1. Start a new post at https://wordpress.com/post
1. Add a poem to the body of the post
1. Save and publish
1. View the post in the Reader stream it https://wordpress.com/

<strong>Result:</strong> line breaks are not displayed for excerpts in summary feeds. This causes a problem for poetry formatting.

screen shot 2015-11-24 at tue nov 24 4 47 29 pm
<sup>Seen at https://wordpress.com/ on this blog https://wordpress.com/read/blog/feed/40471243 for this post https://wordpress.com/read/post/feed/40471243/872344084 using Firefox 42.0 on Mac OS X 10.10.5</sup>

Sample content:

```
why did you go
little fourpaws?
you forgot to shut
your big eyes.

where did you go?
like little kittens

e e cummings
``` 

/hat tip alexjgustafson for the original report

### PR DESCRIPTION
**Pull request by bluefuton**
When displaying excerpts for standard wordpress.com feeds, we use `post.content` from the post object. Currently, we use `post.excerpt` for wordpress.com feeds set to 'summary only'. This means that all tags are stripped, and users have complained that linebreaks are lost (e.g. for poetry):

4a85ec94-92cb-11e5-8849-e31119e585a9

This PR switches the PostExcerpt component to use `post.content` too, and strips tags except `p` and `br`.

We've explored backend solutions for this, such as the `enhanced_excerpt` plugin, but this seems a simpler fix. `post.content` has already been through the post normalizer, and it's used for all other excerpts already.

Fixes #696.

### To test
Pull up a feed set to summary only, and ensure that the excerpt is shown correctly. Examples:
http://calypso.localhost:3000/read/blog/feed/40474296
http://calypso.localhost:3000/read/blog/feed/38960801

screen shot 2016-02-22 at 15 01 49

### CODE REVIEW

File: client/components/post-excerpt/index.jsx [hunk 1]
```diff
@@ -1,31 +1,33 @@
...
+	render() {
+		if ( ! this.props.content ) {
 			return null;
 		}
 
-		if ( text.length > 80 ) {
-			textClass.push( 'is-long' );
-		}
+		// Strip all non-linebreak tags
+		const preparedContent = striptags( this.props.content, [ 'p', 'br' ] );
```
Comment (blowery) [inline client/components/post-excerpt/index.jsx]: let's to this as a part of state. otherwise we're stripping the content on every single render call

Comment (blowery) [inline client/components/post-excerpt/index.jsx]: or move it up into the post normalizer perhaps.

**Comment by blowery** [comment]
Occasionally I'm seeing some oddities when the first paragraph ends up empty:

screenshot

from http://calypso.localhost:3000/read/blog/feed/34575

**Comment by blowery** [comment]
If we do this in post-norm, maybe we walk the stripped DOM and remove any `p` tags that have no children?

**Comment by blowery** [comment]
and if the stripped variant of post.content is empty, fall back to the excerpt?

**Comment by bluefuton** [comment]
> If we do this in post-norm, maybe we walk the stripped DOM and remove any p tags that have no children?

I've done this, but I wonder if it's just the first empty `p` we want to strip. Some excerpts look a bit compressed with all the empty paras gone:

screen shot 2016-02-22 at 18 15 21

**Comment by bluefuton** [comment]
> and if the stripped variant of post.content is empty, fall back to the excerpt?

:+1:

File: client/components/post-excerpt/index.jsx (hunk 1 above)
Comment (bluefuton) [inline client/components/post-excerpt/index.jsx]: :+1: Moved to post-normalizer in c18960e.

**Comment by bluefuton** [comment]
I'll add some post-normalizer tests for this before it goes back up for review.

**Comment by blowery** [comment]
> I've done this, but I wonder if it's just the first empty p we want to strip.

That would be fine, just strip `p`s until we hit one with content.

**Comment by bluefuton** [comment]
> That would be fine, just strip ps until we hit one with content.

:+1: Added in 8c4182d.

File: client/lib/post-normalizer/index.js [hunk 2]
```diff
@@ -469,6 +471,32 @@ normalizePost.withContentDOM = function( transforms ) {
 
 normalizePost.content = {
 
+	createContentWithLinebreakElementsOnly: function createContentWithLinebreakElementsOnly( post, callback ) {
+		if ( ! post.__contentDOM ) {
+			throw new Error( 'this transform must be used as part of withContentDOM' );
+		}
+
+		post.content_with_linebreak_elements_only = striptags( post.__contentDOM.innerHTML, [ 'p', 'br', 'noscript' ] );
```
Comment (blowery) [inline client/lib/post-normalizer/index.js]: why are we keeping `noscript`?

File: client/lib/post-normalizer/index.js [hunk 3]
```diff
@@ -469,6 +471,32 @@ normalizePost.withContentDOM = function( transforms ) {
 
 normalizePost.content = {
 
+	createContentWithLinebreakElementsOnly: function createContentWithLinebreakElementsOnly( post, callback ) {
+		if ( ! post.__contentDOM ) {
```
Comment (blowery) [inline client/lib/post-normalizer/index.js]: this doesn't need to be a content transform if we're not using the contentDOM spun up by the wrapper

File: client/lib/post-normalizer/test/post-normalizer-test.js [hunk 4]
```diff
@@ -870,5 +870,18 @@ describe( 'post-normalizer', function() {
...
 		} );
+		it( 'prepares a version of the post content with linebreak elements only', function( done ) {
+			normalizer(
+				{
+					content: '<p><img src="http://example.com/image.jpg"></p><p><a href="http://wikipedia.org">Giraffes</a> are <br>great</p><p></p>',
+				},
+				[
+					normalizer.withContentDOM( [ normalizer.content.createContentWithLinebreakElementsOnly ] )
+				], function( err, normalized ) {
+					assert.strictEqual( normalized.content_with_linebreak_elements_only, '<p>Giraffes are <br>great</p><p></p>' );
```
Comment (blowery) [inline client/lib/post-normalizer/test/post-normalizer-test.js]: hrm. i think we want to strip all of the empty `p`s? Just feels odd to have the empty ones at the end. Why do we keep 'em?

**Comment by blowery** [comment]
bah, busted things. I'll fix.

File: client/lib/post-normalizer/index.js (hunk 2 above)
Comment (bluefuton) [inline client/lib/post-normalizer/index.js]: Yep. I think we can drop this now we're running `createContentWithLinebreakElementsOnly` after `detectPolls` in the post normalizer.

**Comment by bluefuton** [comment]
Todo:
- [ ] strip empty `<br />` from beginning of content
- [ ] strip any element with the class `wp-caption-text` from beginning of content (see https://cloudup.com/cYJAs2-taRl/f)

**Comment by bluefuton** [comment]
Is `createContentWithLinebreakElementsOnly` too verbose? Would `createBetterExcerpt` do it?

**Comment by blowery** [comment]
> Is createContentWithLinebreakElementsOnly too verbose? Would createBetterExcerpt do it?

Maybe `createEnhancedExcerpt`?

**Comment by bluefuton** [comment]
> Maybe createEnhancedExcerpt?

I thought that, but probably want to avoid confusion with the server-side `enhanced_excerpt` plugin (which retains much more of the content including embeds and images).

**Comment by bluefuton** [comment]
Naming aside, this should be ready to roll now.

**Pull request merged**